core/liftRecursive.ml

```diff
@@ -0,0 +1,30 @@
+open Sugartypes
+open Utility
+open SourceCode
+
+(* A simple pass to lift (self)-recursive functions from a `Fun` node to a `Funs` node. *)
+
+(* Simple check to see whether a function is recursive: checks whether the
+ * binder occurs within the free variables of the function body. *)
+(* Note that we do not need to handle mutually-recursive functions
+ * here since they will be constructed explicitly during parsing. *)
+let is_recursive bnd fnlit =
+  StringSet.mem (Binder.to_name bnd) (Freevars.funlit fnlit)
```

Comment:
In OCaml (and Scheme) there is a distinction between `let` and `letrec`, but in Standard ML (and probably most languages) there is only one function definition construct and it is implicitly recursive.  In the absence of such a distinction I think it's OK for the default to be recursive.  One can fake `let` by doing `var f = fun (...) {...}` if one wants to name a non-recursive function that refers to a previously defined function of the same name, if one really wants this.

I think here, all that changes is that previously two consecutive `fun` definitions would not be allowed at all, without intervening non-function bindings, and if separated by something else each one would have the same meaning as now.  So I'm OK with this implication because it does not change the behavior of existing working code.

Reply:
Good observation. I don't think it's problematic enough to go with non-recursive functions by default (as I say, the ripple would be more like a tidal wave), but it's worth knowing about (and would be a good warning candidate). Thanks.

Reply:
okay, I'm perfectly fine with keeping things as they are

Reply:
In `typeSugar.ml` when typechecking `Fun` there is a hack:

```
(* HACK: Place a dummy name in the environment in
   order to ensure that the generalisation check
   does the right thing (it would be unsound to use
   the original name as the function is not
   recursive) *)
let v = Utils.dummy_source_name () in
bind_var context (v, fti), ft
```
It is still necessary given that the functions are recursive? I tried using `bind_var context (Binder.to_name bndr, fti), ft` instead and tests don't break.

Reply:
I'd like to point out to the above question to make sure it doesn't go unnoticed.

Reply:
Sorry, I missed this. After the `liftFunctions` pass, all recursive functions are promoted to singleton `Funs` blocks, so we can assume that all `Fun` functions are explicitly non-recursive.

I wouldn't be able to give you a specific answer to your question (maybe @dhil, @frank-emrich, @jamescheney, or @slindley would be more familiar with this part of the typechecker?)

Reply:
In the HM generalization step, you need the current variable environment because you don't want to generalize type variables that already occur in the variable environment (i.e., in the type of any of the variables bound in the current context).
I'm quite sure that this is why we add a dummy variable to the context here, whose type is the type of the current function. However, I don't understand the exact reason why this is needed here...
In any case, I think that this PR doesn't change anything this piece of code relies on, in particular since `Fun` and `Funs` should be used in the same way as they were before.